Source/WebKit/Platform/cocoa/WebPrivacyHelpers.mm

```diff
@@ -239,6 +249,81 @@ void requestLinkDecorationFilteringData(LinkFilteringRulesCallback&& callback)
     }];
 }
 
+StorageAccessPromptQuirkController& StorageAccessPromptQuirkController::shared()
+{
+    static StorageAccessPromptQuirkController* sharedInstance = new StorageAccessPromptQuirkController;
+    return *sharedInstance;
+}
+
+Ref<StorageAccessPromptQuirkObserver> StorageAccessPromptQuirkController::observeUpdates(Function<void()>&& completionHandler)
+{
+    ASSERT(RunLoop::isMain());
+    if (!m_notificationListener) {
+        m_notificationListener = adoptNS([WKWebPrivacyNotificationListener new]);
+        [m_notificationListener listenForStorageAccessPromptQuirkChanges:^{
+            updateQuirks([this] {
+                for (auto& observer : m_observers)
+                    observer.invokeCallback();
+            });
+        }];
+    }
+    Ref observer = StorageAccessPromptQuirkObserver::create(WTFMove(completionHandler));
+    m_observers.add(observer.get());
+    return observer;
+}
+
+void StorageAccessPromptQuirkController::setCachedQuirks(Vector<WebCore::OrganizationStorageAccessPromptQuirk>&& quirks)
+{
+    m_cachedQuirks = WTFMove(quirks);
+    m_cachedQuirks.shrinkToFit();
+}
+
+static HashMap<WebCore::RegistrableDomain, Vector<WebCore::RegistrableDomain>> domainPairingsDictToMap(NSDictionary<NSString *, NSArray<NSString *> *> *domainPairings)
+{
+    HashMap<WebCore::RegistrableDomain, Vector<WebCore::RegistrableDomain>> map;
+    auto* topDomains = domainPairings.allKeys;
+    for (NSString *topDomain : topDomains) {
+        Vector<WebCore::RegistrableDomain> subFrameDomains;
+        for (NSString *subFrameDomain : [domainPairings objectForKey:topDomain])
+            subFrameDomains.append(WebCore::RegistrableDomain::fromRawString(subFrameDomain));
+        map.set(WebCore::RegistrableDomain::fromRawString(String { topDomain }), WTFMove(subFrameDomains));
+    }
+    return map;
+}
+
+void StorageAccessPromptQuirkController::updateQuirks(CompletionHandler<void()>&& completionHandler)
+{
+    ASSERT(RunLoop::isMain());
+    if (!WebKit::canUseWebPrivacyFramework()) {
+        completionHandler();
+        return;
+    }
+
+    static MainThreadNeverDestroyed<Vector<CompletionHandler<void()>, 1>> lookupCompletionHandler;
+    lookupCompletionHandler->append(WTFMove(completionHandler));
+    if (lookupCompletionHandler->size() > 1)
+        return;
+
+    RetainPtr options = adoptNS([PAL::allocWPResourceRequestOptionsInstance() init]);
+    [options setAfterUpdates:NO];
+
+    [[PAL::getWPResourcesClass() sharedInstance] requestStorageAccessPromptQuirksData:options.get() completionHandler:^(WPStorageAccessPromptQuirksData *data, NSError *error) {
```

Comment:
This'll need a `-respondsToSelector:` check to avoid crashing, if the changes in WP haven't been submitted yet.